MSC XXXX: Add supported appservice version to registration information (#2379)
matrix-org/matrix-doc: proposals/2379-appservice-versions.md

# MSC 2379: Add /versions endpoint to Appservice API. 

Bridges do not have a way to specify what version of the spec they support. This means that if the path
of any of the appservice endpoints were to change in the spec, homeservers would not be able to
intelligently discover the paths that a bridge supports.

## Proposal

A new endpoint is required, which is `/_matrix/app/versions`. This is nearly identical to the
[C-S API](https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-versions) endpoint
but lacks a `unstable_features` key, and is hosted by the appservice rather than the homeserver.

All bridges MUST implement this endpoint and specify which version(s) of the `AS` API they support. 

> The homeserver MUST send requests to the endpoints specified by that version of the AS spec.

Reviewer: Last thing I think I'd put in is a Legacy Considerations section that states if an appservice hasn't implemented `/versions`, a homeserver can choose to either assume it supports all AS API versions up until the one where this `/versions` endpoint was introduced, or just deny connecting to it at all.

Author: ...but that's what I just deleted :cry:

Reviewer: I wanted the stuff about specifying new endpoints deleted, but this bit we should keep.

# MSC 2379: Add supported appservice version to registration information. 

The [AS registration format](https://matrix.org/docs/spec/application_service/r0.1.2#registration) does
not have a way to specify what version of the spec that bridges support. This means that if the path
of any of the appservice endpoints were to change in the spec, homeservers would not be able to
intelligently discover the paths that a bridge supports.

## Proposal

> The `registration` file should contain one new key: `as_version`.

Reviewer 2: Overall I'm not convinced we need this in the spec. Synapse already supports arbitrary properties on the registration file, and the registration file in the spec is not a super strict thing (like anything, it's extensible). It doesn't even have to be YAML.

Author: Sure, but this isn't just about Synapse. This is about ensuring that bridges have a single way to tell any homeserver impl which paths they should be hitting, rather than each homeserver specifying their own way of doing this.

Right now any homeserver that wishes to implement the appservice API is not going to be following the spec anyway, because at least the matrix-org bridges do not implement the correct endpoints. Presumably there are other bridges out there too that may not have bothered to follow the spec, because synapse doesn't use it. Dendrite is also non-compliant, for ex.

Yes, we could fix the bug we have right now for *Synapse* onlu by ensuring that every bridge developer out there appends a synapse-only key to their registration format (or fixes their shit), but that seems a like a waste of time if we have this problem with other homeservers down the line.

Furthermore, this change also future proof the API when we inevitably change the format of the transaction format.

The value of the key should be set to the AS API version (`rX.X.X`) that the bridge supports. This
key SHOULD be specified by all bridges. When set, the homeserver MUST send requests to the endpoints
specified by that version of the AS spec.

> Homeservers may optionally support an omitted value, which will make it support the legacy paths used

Author: This stuff is necessary so that we do not kill all bridges everywhere if we start supporting this, but it does suck from a spec standpoint of having to include it. Perhaps we don't need to spec this and leave it as an undocumented synapse feature to be removed at some point in the future?

by Synapse `<=1.6.X`.

> The legacy paths omit the /_matrix/app/{version} prefix entirely for:

Reviewer 2: ```suggestion
The legacy paths omit the `/_matrix/app/{version}` prefix entirely for:
```

[C-S API](https://matrix.org/docs/spec/client_server/r0.6.0#get-matrix-client-versions) endpoint
but lacks a `unstable_features` key, and is hosted by the appservice rather than the homeserver.

All bridges SHOULD implement this endpoint and specify which version(s) of the `AS` API they support. 
The homeserver MUST send requests to the endpoints specified by that version of the AS spec.

Homeservers may optionally support a 404 response to this endpoint, which will make it use the legacy paths used
by Synapse `<=1.6.X`. Alternatively, the homeserver may choose not to communicate with a bridge that lacks a /versions
endpoint.

The legacy paths omit the /_matrix/app/{version} prefix entirely for:

    - `/_matrix/app/{version}/transactions/{txnId}` becomes  `/transactions/{txnId}`
    - `/_matrix/app/{version}/users/{userId}` becomes  `/users/{userId}`
    - `/_matrix/app/{version}/rooms/{roomAlias}` becomes  `/rooms/{roomAlias}`

Additionally, the `{version}` for the Third party network routes is always set to `unstable`.

It should be reiterated that support for this is up to the homeserver implemetor. Homeservers may

> refuse to load appservices that do not include this `key`.

Reviewer: ```suggestion
refuse to load appservices that do not include this endpoint.
```
?

## Potential issues

Keeping a 'legacy' mode around in the spec sucks, because it's horribly non-compliant to the version system.
However, most of the ecosystem has been modeled over Synapse behaviours which means this spec change would break
support for bridges if implemented by Synapse. This option remains the most pragmatic option. In a future version

> of the spec, this mode could be removed.

Reviewer: ```suggestion
of the spec, this legacy behaviour could be removed.
```
